src/logging.c

```diff
@@ -0,0 +1,151 @@
+// Copyright 2017 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#if __cplusplus
+extern "C"
+{
+#endif
+
+#include "rcutils/allocator.h"
+#include "rcutils/logging.h"
+
+bool g_rcl_logging_initialized = false;
+
+rcl_logging_output_handler_t g_rcl_logging_output_handler = NULL;
+
+int g_rcl_logging_severity_threshold = 0;
+
+void rcl_logging_initialize()
+{
+  if (!g_rcl_logging_initialized) {
+    g_rcl_logging_output_handler = &rcl_logging_console_output_handler;
+    g_rcl_logging_severity_threshold = RCUTILS_LOG_SEVERITY_DEBUG;
+    g_rcl_logging_initialized = true;
+  }
+}
+
+rcl_logging_output_handler_t rcl_logging_get_output_handler()
+{
+  return g_rcl_logging_output_handler;
+}
+
+void rcl_logging_set_output_handler(rcl_logging_output_handler_t function)
+{
+  g_rcl_logging_output_handler = function;
+}
+
+int rcl_logging_get_severity_threshold()
+{
+  return g_rcl_logging_severity_threshold;
+}
+
+void rcl_logging_set_severity_threshold(int severity)
+{
+  g_rcl_logging_severity_threshold = severity;
+}
+
+void rcl_log(
+  rcl_log_location_t * location,
+  int severity, const char * name, const char * format, ...)
+{
+  if (severity < g_rcl_logging_severity_threshold) {
+    return;
+  }
+  rcl_logging_output_handler_t output_handler = g_rcl_logging_output_handler;
+  if (output_handler) {
+    va_list args;
+    va_start(args, format);
+    (*output_handler)(location, severity, name, format, &args);
+    va_end(args);
+  }
+}
+
+void rcl_logging_console_output_handler(
+  rcl_log_location_t * location,
+  int severity, const char * name, const char * format, va_list * args)
+{
```

Comment:
Would it make sense to make severity an "enum RCUTILS_LOG_SEVERITY" instead of an int (here and elsewhere)?  It won't make a huge difference to the code, but would make it clear to anyone implementing their own output handler what the types they can expect are.